File: src/lib/lautils.cpp
#include "lautils.h"
#include "laanimal.h"
#include "lacrop.h"
#include "laanimalparameter.h"
#include "lacropparameter.h"
#include "la.h"

#include <QApplication>
#include <QDir>
#include <QFile>
#include <QPluginLoader>
#include <QSettings>
#include <QString>
#include <QStringList>
#include <QVector>
#include <QtXml>

#include <QFileDialog>
#ifdef Q_OS_MACX
//for getting app bundle path
#include <ApplicationServices/ApplicationServices.h>
#endif


/**
 * Returns the path to the settings directory in user's home dir
 */
const QString LaUtils::userSettingsDirPath()
{
  QSettings mySettings;
  QString myPath=
      mySettings.value("dataDirs/dataDir", QDir::homePath() + QString("/.landuseAnalyst/") ).toString();
  //  Make sure the users settings dir actually exists
  //qDebug("LaUtils::userSettingsDirPath() = " + myPath.toLocal8Bit());
  QDir().mkpath(myPath);
  return myPath;
}
const QString LaUtils::getModelOutputDir()
{
  QString myPath = userSettingsDirPath()+QDir::separator()+"modelOutputs"+QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}
const QString LaUtils::userAnimalProfilesDirPath()
{
  //alg profiles are always saved in the users home dir under .landuseAnalyst/
  QString myPath = QDir::homePath() + QString("/.landuseAnalyst/") +
    QDir::separator()+"animalProfiles"+QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}


const QString LaUtils::userCropProfilesDirPath()
{
  //alg profiles are always saved in the users home dir under .landuseAnalyst/
  QString myPath = QDir::homePath() + QString("/.landuseAnalyst/") +
    QDir::separator()+"cropProfiles"+QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}


LaUtils::AnimalMap LaUtils::getAvailableAnimals()
{
  LaUtils::AnimalMap myMap;
  QDir myDirectory(userAnimalProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading animal: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaAnimal myAnimal;
      myAnimal.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myAnimal.name().isEmpty())
      {
        qDebug("Animal name was empty!");
        continue;
      }
      //qDebug("Adding " + myAnimal.name());
      //qDebug(myAnimal.toText().toLocal8Bit());
      myMap[myAnimal.guid()]=myAnimal;
    }
  }
  return myMap;
}
LaAnimal LaUtils::getAnimal(QString theGuid)
{
  QDir myDirectory(userAnimalProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading animal: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaAnimal myAnimal;
      myAnimal.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myAnimal.name().isEmpty())
      {
        qDebug("Animal name was empty!");
        continue;
      }
      //qDebug("Adding " + myAnimal.name());
      //qDebug(myAnimal.toText().toLocal8Bit());
      if (myAnimal.guid()==theGuid)
      {
        return myAnimal;
      }
    }
  }
  LaAnimal myAnimal; //blank animal
  return myAnimal;
}

LaUtils::CropMap LaUtils::getAvailableCrops()
{
  LaUtils::CropMap myMap;
  QDir myDirectory(userCropProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading crop: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaCrop myCrop;
      myCrop.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myCrop.name().isEmpty())
      {
        qDebug("Crop name was empty!");
        continue;
      }
      //qDebug("Adding " + myCrop.name());
      myMap[myCrop.guid()]=myCrop;
      //qDebug(myCrop.toText().toLocal8Bit());
    }
  }
  return myMap;
}

LaCrop LaUtils::getCrop(QString theGuid)
{
  QDir myDirectory(userCropProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading crop: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaCrop myCrop;
      myCrop.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myCrop.name().isEmpty())
      {
        qDebug("Crop name was empty!");
        continue;
      }
      //qDebug("Adding " + myCrop.name());
      //qDebug(myCrop.toText().toLocal8Bit());
      if (myCrop.guid()==theGuid)
      {
        return myCrop;
      }
    }
  }
  LaCrop myCrop; //blank crop
  return myCrop;
}
const QString LaUtils::userConversionTablesDirPath()
{
  // always saved in the users home dir under .landuseAnalyst/
  QString myPath = QDir::homePath() + QString("/.landuseAnalyst/") +
    QDir::separator()+"conversionTables"+QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}

const QString LaUtils::userAnimalParameterProfilesDirPath()
{
  //alg profiles are always saved in the users home dir under .landuseAnalyst/
  QString myPath = QDir::homePath() + QString("/.landuseAnalyst/") +
    QDir::separator()+"animalParameterProfiles"+QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}

const QString LaUtils::userImagesDirPath()
{
  QString myPath = QDir::homePath() + QString("/.landuseAnalyst") +
    QDir::separator()+"images"+QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}

const QString LaUtils::userCropParameterProfilesDirPath()
{
  //alg profiles are always saved in the users home dir under .landuseAnalyst/
  QString myPath = QDir::homePath() + QString("/.landuseAnalyst/") +
    QDir::separator() + "cropParameterProfiles" + QDir::separator();
  QDir().mkpath(myPath);
  return myPath;
}

int LaUtils::convertAreaToHectares(AreaUnits theAreaUnit, int theArea)
{
  // this may seem ridiculous to do it this way, but
  // i plan to include other area units and this way
  // it will make it very easy to work with in the future
  // all need be done is add new units to la.h enum and enter
  // into the following switch...

  float myHectares = 0.;

  switch (theAreaUnit)  // Dunum, Hectare
  {
    case  Dunum:
      myHectares = theArea * 10.;
      break;
    case  Hectare:
      myHectares = theArea;
      break;   /*    add new units here after updating la.h enum ex:
    case  Acre:
      HyHectares = theArea / 2.47105381;
      break;     */
  }
  return static_cast<int>(myHectares);
}

LaUtils::AnimalParameterMap LaUtils::getAvailableAnimalParameters()
{
  LaUtils::AnimalParameterMap myMap;
  QDir myDirectory(userAnimalParameterProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading animalParameter: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaAnimalParameter myAnimalParameter;
      myAnimalParameter.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myAnimalParameter.name().isEmpty())
      {
        qDebug("AnimalParameter name was empty!");
        continue;
      }
      //qDebug("Adding " + myAnimalParameter.name());
      //qDebug(myAnimalParameter.toText().toLocal8Bit());
      myMap[myAnimalParameter.guid()]=myAnimalParameter;


      //for debug only...
      ;
      qDebug(" ++ lautil  Restoring " +
          QString::number(myAnimalParameter.fodderSourceMap().count()).toLocal8Bit()
          + " food sources into animal parameter.");
    }
  }
  return myMap;
}

LaAnimalParameter LaUtils::getAnimalParameter(QString theGuid)
{
  QDir myDirectory(userAnimalParameterProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading animalParameter: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaAnimalParameter myAnimalParameter;
      myAnimalParameter.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myAnimalParameter.name().isEmpty())
      {
        qDebug("AnimalParameter name was empty!");
        continue;
      }
      //qDebug("Adding " + myAnimalParameter.name());
      //qDebug(myAnimalParameter.toText().toLocal8Bit());
      if (myAnimalParameter.guid()==theGuid)
      {
        return myAnimalParameter;
      }
    }
  }
  LaAnimalParameter myAnimalParameter;
  return myAnimalParameter; //retrun a blank one since no match found

}

LaUtils::CropParameterMap LaUtils::getAvailableCropParameters()
{
  LaUtils::CropParameterMap myMap;
  QDir myDirectory(userCropParameterProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading cropParameter: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaCropParameter myCropParameter;
      myCropParameter.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myCropParameter.name().isEmpty())
      {
        //qDebug("CropParameter name was empty!");
        continue;
      }
      //qDebug("Adding " + myCropParameter.name());
      myMap[myCropParameter.guid()]=myCropParameter;
      //qDebug(myCropParameter.toText().toLocal8Bit());
    }
  }
  return myMap;
}

LaCropParameter LaUtils::getCropParameter(QString theGuid)
{
  QDir myDirectory(userCropParameterProfilesDirPath());
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //Ignore directories
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if the filename ends in .xml try to load it into our layerSets listing
    if(myFileInfo.completeSuffix()=="xml")
    {
      //qDebug("Loading animalParameter: " + myList.at(i).absoluteFilePath().toLocal8Bit());
      LaCropParameter myCropParameter;
      myCropParameter.fromXmlFile(myFileInfo.absoluteFilePath());
      if (myCropParameter.name().isEmpty())
      {
        qDebug("CropParameter name was empty!");
        continue;
      }
      //qDebug("Adding " + myCropParameter.name());
      //qDebug(myCropParameter.toText().toLocal8Bit());
      if (myCropParameter.guid()==theGuid)
      {
        return myCropParameter;
      }
    }
  }
  LaCropParameter myCropParameter;
  return myCropParameter; //retrun a blank one since no match found

}


QStringList LaUtils::sortList(QStringList theList)
{
    //sort the taxon list alpabetically descending order
    theList.sort(); //this sorts ascending!
    //flip the sort order
    QStringList mySortedList;
    QStringList::Iterator myIterator= theList.end();
    while( myIterator!=theList.begin() )
    {
      myIterator--;
      mySortedList << *myIterator;
    }
    return mySortedList;
}

QStringList LaUtils::uniqueList(QStringList theList)
{
    //remove any duplicates from a sorted list
    QStringList myUniqueList;
    QString myLast = "";
    QStringListIterator myIterator( theList );
    while( myIterator.hasNext() )
    {
      QString myCurrent = myIterator.next();
      if (myCurrent!=myLast)
      {
        myUniqueList << myCurrent;
      }
      myLast=myCurrent;
    }
    return myUniqueList;
}

//return a string list of all the experiment files
QStringList LaUtils::getExperimentsList()
{
  QStringList myExperimentList;
  QSettings mySettings;
  QString myWorkDir = mySettings.value("dataDirs/dataDir",QDir::homePath()+QDir::separator()+".landuseAnalyst").toString()+
                      QString("/modelOutputs/");
  QDir myDirectory(myWorkDir);
  myDirectory.setFilter(QDir::Dirs | QDir::Files | QDir::NoSymLinks );
  //qDebug ("Current directory is: " +  myWorkDir.toAscii());
  QFileInfoList myList = myDirectory.entryInfoList();
  for (unsigned int i = 0; i < static_cast<unsigned int>(myList.size()); ++i)
  {
    QFileInfo myFileInfo = myList.at(i);
    //qDebug("Get ExperimentsList Scanning : " + myFileInfo.fileName().toLocal8Bit());
    if(myFileInfo.fileName() == "." ||myFileInfo.fileName() == ".." )
    {
      continue;
    }
    //if this is an empty dir, just skip it...
    if (myFileInfo.absoluteDir().count() < 1)
    {
      continue;
    }
    //todo - check that the file is a directory (it should be)

    //Now look in each dir under the workdir for the xml experiment file
    //the experiment file should be named <dir name>.xml
    QString myFile = myFileInfo.absolutePath() + QDir::separator()
                                 + myFileInfo.fileName() + QDir::separator()
                                 + myFileInfo.fileName() + ".xml";
    //qDebug("Get ExperimentsList Checking: " + myFile.toLocal8Bit());
    if (QFile::exists(myFile))
    {
      myExperimentList << myFile;
      //qDebug("Get ExperimentsList Adding: " + myFile.toLocal8Bit());
    }
  }
  return myExperimentList;
}

bool LaUtils::createTextFile(QString theFileName, QString theData)
{
  //create the txt file
  QFile myFile( theFileName );
  if ( myFile.open( QIODevice::WriteOnly ) )
  {
    QTextStream myQTextStream( &myFile );
    myQTextStream << theData;
  }
  else
  {
    return false;
  }
  myFile.close();
  return true ;
}

QString LaUtils::xmlEncode(QString theString)
{
  theString = theString.replace("<","&lt;");
  theString = theString.replace(">","&gt;");
  theString = theString.replace("&","&amp;");
  return theString;
}

QString LaUtils::xmlDecode(QString theString)
{
  theString = theString.replace("&lt;","<");
  theString = theString.replace("&gt;",">");
  theString = theString.replace("&amp;","&");
  return theString;
}

QString LaUtils::getStandardCss()
{
  QString myStyle = ".glossy{ background-color: qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 #616161, stop: 0.5 #505050, stop: 0.6 #434343, stop:1 #656565); color: white; padding-left: 4px; border: 1px solid #6c6c6c; }";
  myStyle += "body {background: white;}";
  myStyle += "h1 {font-size : 22pt; color: #0063F7; }";
  myStyle += "h2 {font-size : 18pt; color: #0063F7; }";
  myStyle += "h3 {font-size : 14pt; color: #0063F7; }";
  myStyle += ".cellHeader {color:#466aa5; font-size : 12pt;}";
  myStyle += ".parameterHeader {font-weight: bold;}";
  myStyle += ".largeCell {color:#000000; font-size : 12pt;}";
  myStyle += ".table {"
                    "  border-width: 1px 1px 1px 1px;"
                    "  border-spacing: 2px;"
                    "  border-style: solid solid solid solid;"
                    "  border-color: black black black black;"
                    "  border-collapse: separate;"
                    "  background-color: white;"
                    "}";
  return myStyle;
}

QString LaUtils::openGraphicFile()
{
  QString myHomePath = QDir::homePath();
  QString myFileName = QFileDialog::getOpenFileName(0, "Choose an image", myHomePath, "Images (*.png *.xpm *.jpg)");
  QFileInfo fi(myFileName);
  QString myName = fi.fileName();
  QString myDestinationFilePathName = userImagesDirPath() + myName;
  QFile::copy(myFileName, myDestinationFilePathName);
  return myDestinationFilePathName;
}

QString LaUtils::saveFile()
{
  QString myHomePath = QDir::homePath();
  QString myFileName = QFileDialog::getSaveFileName(0, "Choose a file name", userConversionTablesDirPath(), "*.csv");
  QFileInfo fi(myFileName);
  QString myName = fi.fileName();
  QString myDestinationFilePathName = userConversionTablesDirPath() + myName;
  return myDestinationFilePathName;
}

File: landuse_analyst/lib/lautils.py
from qgis.PyQt.QtCore import QDir, QFile, QIODevice, QSettings, QTextStream, QFileInfo
from qgis.PyQt.QtWidgets import QFileDialog
from typing import Dict, List
import os
import xml.etree.ElementTree as ET

from landuse_analyst.lib.laanimal import LaAnimal
from landuse_analyst.lib.lacrop import LaCrop
from landuse_analyst.lib.laanimalparameter import LaAnimalParameter
from landuse_analyst.lib.lacropparameter import LaCropParameter

class LaUtils:
    @staticmethod
    def openGraphicFile(parent=None):
        """
        Opens a file dialog to select a graphic file and returns the file path.
        """
        print("openGraphicFile called")
        options = QFileDialog.Options()
        file_path, _ = QFileDialog.getOpenFileName(
            parent,
            "Open Graphic File",
            QDir.homePath(),
            "Images (*.png *.xpm *.jpg *.jpeg *.bmp *.gif)",
            options=options
        )
        print(f"Selected file path: {file_path}")
        return file_path

    @staticmethod
    def userSettingsDirPath():
        """
        Returns the path to the settings directory in user's home dir.
        """
        settings = QSettings()
        path = settings.value("dataDirs/dataDir", QDir.homePath() + "/.landuseAnalyst/")
        QDir().mkpath(path)
        return path

    @staticmethod
    def getModelOutputDir():
        """
        Returns the path to the model output directory.
        """
        path = LaUtils.userSettingsDirPath() + QDir.separator() + "modelOutputs" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def userAnimalProfilesDirPath():
        """
        Returns the path to the user animal profiles directory.
        """
        path = QDir.homePath() + "/.landuseAnalyst/" + QDir.separator() + "animalProfiles" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def userCropProfilesDirPath():
        """
        Returns the path to the user crop profiles directory.
        """
        path = QDir.homePath() + "/.landuseAnalyst/" + QDir.separator() + "cropProfiles" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def getAvailableCrops() -> Dict[str, LaCrop]:
        """
        Returns a dictionary of available crops.
        """
        # Initialize an empty dictionary to store the crops
        crop_map = {}

        # Get the path to the user crop profiles directory
        directory = QDir(LaUtils.userCropProfilesDirPath())
        print(f"Crop profiles directory: {directory.absolutePath()}")

        # Set the filter to include directories and files, but exclude symbolic links
        directory.setFilter(QDir.Dirs | QDir.Files | QDir.NoSymLinks)

        # Get the list of files and directories in the crop profiles directory
        file_list = directory.entryInfoList()
        print(f"Number of files found: {len(file_list)}")

        # Iterate through the list of files and directories
        for file_info in file_list:
            print(f"Processing file: {file_info.fileName()}")

            # Ignore the current directory (.) and parent directory (..)
            if file_info.fileName() in [".", ".."]:
                print("Ignoring current or parent directory")
                continue

            # If the file has an XML extension, try to load it as a crop profile
            if file_info.completeSuffix() == "xml":
                print(f"Found XML file: {file_info.fileName()}")
                # Create a LaCrop object and load the crop profile from the XML file
                crop = LaCrop()
                crop.fromXmlFile(file_info.absoluteFilePath())

                # If the crop name is not empty, add it to the dictionary
                if crop.name():
                    print(f"Loaded crop: {crop.name()} with GUID: {crop.guid()}")
                    crop_map[crop.guid()] = crop
                else:
                    print("Crop name is empty, skipping file")
            else:
                print(f"Skipping non-XML file: {file_info.fileName()}")

        # Return the dictionary of available crops
        print(f"Total crops loaded: {len(crop_map)}")
        return crop_map

    @staticmethod
    def userConversionTablesDirPath():
        """
        Returns the path to the user conversion tables directory.
        """
        path = QDir.homePath() + "/.landuseAnalyst/" + QDir.separator() + "conversionTables" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def userAnimalParameterProfilesDirPath():
        """
        Returns the path to the user animal parameter profiles directory.
        """
        path = QDir.homePath() + "/.landuseAnalyst/" + QDir.separator() + "animalParameterProfiles" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def userImagesDirPath():
        """
        Returns the path to the user images directory.
        """
        path = QDir.homePath() + "/.landuseAnalyst" + QDir.separator() + "images" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def userCropParameterProfilesDirPath():
        """
        Returns the path to the user crop parameter profiles directory.
        """
        path = QDir.homePath() + "/.landuseAnalyst/" + QDir.separator() + "cropParameterProfiles" + QDir.separator()
        QDir().mkpath(path)
        return path

    @staticmethod
    def convertAreaToHectares(area_unit: int, area: int) -> int:
        """
        Converts area to hectares.
        """
        hectares = 0.0
        if area_unit == 0:  # Dunum
            hectares = area * 0.1
        elif area_unit == 1:  # Hectare
            hectares = area
        return int(hectares)

    @staticmethod
    def getAvailableAnimalParameters() -> Dict[str, LaAnimalParameter]:
        """
        Returns a dictionary of available animal parameters.
        """
        animal_parameter_map = {}
        directory = QDir(LaUtils.userAnimalParameterProfilesDirPath())
        directory.setFilter(QDir.Dirs | QDir.Files | QDir.NoSymLinks)
        file_list = directory.entryInfoList()
        for file_info in file_list:
            if file_info.fileName() in [".", ".."]:
                continue
            if file_info.completeSuffix() == "xml":
                animal_parameter = LaAnimalParameter()
                animal_parameter.fromXmlFile(file_info.absoluteFilePath())
                if animal_parameter.name():
                    animal_parameter_map[animal_parameter.guid()] = animal_parameter
        return animal_parameter_map

    @staticmethod
    def getAnimalParameter(guid: str) -> LaAnimalParameter:
        """
        Returns an animal parameter by its GUID.
        """
        directory = QDir(LaUtils.userAnimalParameterProfilesDirPath())
        directory.setFilter(QDir.Dirs | QDir.Files | QDir.NoSymLinks)
        file_list = directory.entryInfoList()
        for file_info in file_list:
            if file_info.fileName() in [".", ".."]:
                continue
            if file_info.completeSuffix() == "xml":
                animal_parameter = LaAnimalParameter()
                animal_parameter.fromXmlFile(file_info.absoluteFilePath())
                if animal_parameter.name() and animal_parameter.guid() == guid:
                    return animal_parameter
        return LaAnimalParameter()  # Return a blank animal parameter if no match found

    @staticmethod
    def getAvailableCropParameters() -> Dict[str, LaCropParameter]:
        """
        Returns a dictionary of available crop parameters.
        """
        crop_parameter_map = {}
        directory = QDir(LaUtils.userCropParameterProfilesDirPath())
        directory.setFilter(QDir.Dirs | QDir.Files | QDir.NoSymLinks)
        file_list = directory.entryInfoList()
        for file_info in file_list:
            if file_info.fileName() in [".", ".."]:
                continue
            if file_info.completeSuffix() == "xml":
                crop_parameter = LaCropParameter()
                crop_parameter.fromXmlFile(file_info.absoluteFilePath())
                if crop_parameter.name():
                    crop_parameter_map[crop_parameter.guid()] = crop_parameter
        return crop_parameter_map

    @staticmethod
    def getCropParameter(guid: str) -> LaCropParameter:
        """
        Returns a crop parameter by its GUID.
        """
        directory = QDir(LaUtils.userCropParameterProfilesDirPath())
        directory.setFilter(QDir.Dirs | QDir.Files | QDir.NoSymLinks)
        file_list = directory.entryInfoList()
        for file_info in file_list:
            if file_info.fileName() in [".", ".."]:
                continue
            if file_info.completeSuffix() == "xml":
                crop_parameter = LaCropParameter()
                crop_parameter.fromXmlFile(file_info.absoluteFilePath())
                if crop_parameter.name() and crop_parameter.guid() == guid:
                    return crop_parameter
        return LaCropParameter()  # Return a blank crop parameter if no match found

    @staticmethod
    def sortList(the_list: List[str]) -> List[str]:
        """
        Sorts the list alphabetically in descending order.
        """
        the_list.sort()
        return the_list[::-1]

    @staticmethod
    def uniqueList(the_list: List[str]) -> List[str]:
        """
        Removes duplicates from a sorted list.
        """
        unique_list = []
        last = ""
        for current in the_list:
            if current != last:
                unique_list.append(current)
            last = current
        return unique_list

    @staticmethod
    def getExperimentsList() -> List[str]:
        """
        Returns a list of all the experiment files.
        """
        experiment_list = []
        settings = QSettings()
        work_dir = settings.value("dataDirs/dataDir", QDir.homePath() + QDir.separator() + ".landuseAnalyst") + "/modelOutputs/"
        directory = QDir(work_dir)
        directory.setFilter(QDir.Dirs | QDir.Files | QDir.NoSymLinks)
        file_list = directory.entryInfoList()
        for file_info in file_list:
            if file_info.fileName() in [".", ".."]:
                continue
            if file_info.absoluteDir().count() < 1:
                continue
            experiment_file = file_info.absolutePath() + QDir.separator() + file_info.fileName() + QDir.separator() + file_info.fileName() + ".xml"
            if QFile.exists(experiment_file):
                experiment_list.append(experiment_file)
        return experiment_list

    @staticmethod
    def createTextFile(file_name: str, data: str) -> bool:
        """
        Creates a text file with the given data.
        """
        file = QFile(file_name)
        if file.open(QIODevice.WriteOnly):
            text_stream = QTextStream(file)
            text_stream << data
            file.close()
            return True
        return False

    @staticmethod
    def xmlEncode(string: str) -> str:
        """
        Encodes a string for XML.
        """
        string = string.replace("<", "&lt;")
        string = string.replace(">", "&gt;")
        string = string.replace("&", "&amp;")
        return string

    @staticmethod
    def xmlDecode(string: str) -> str:
        """
        Decodes a string from XML.
        """
        string = string.replace("&lt;", "<")
        string = string.replace("&gt;", ">")
        string = string.replace("&amp;", "&")
        return string

    @staticmethod
    def getStandardCss() -> str:
        """
        Returns the standard CSS.
        """
        style = ".glossy{ background-color: qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 #616161, stop: 0.5 #505050, stop: 0.6 #434343, stop:1 #656565); color: white; padding-left: 4px; border: 1px solid #6c6c6c; }"
        style += "body {background: white;}"
        style += "h1 {font-size : 22pt; color: #0063F7; }"
        style += "h2 {font-size : 18pt; color: #0063F7; }"
        style += "h3 {font-size : 14pt; color: #0063F7; }"
        style += ".cellHeader {color:#466aa5; font-size : 12pt;}"
        style += ".parameterHeader {font-weight: bold;}"
        style += ".largeCell {color:#000000; font-size : 12pt;}"
        style += ".table {"
        style += "  border-width: 1px 1px 1px 1px;"
        style += "  border-spacing: 2px;"
        style += "  border-style: solid solid solid solid;"
        style += "  border-color: black black black black;"
        style += "  border-collapse: separate;"
        style += "  background-color: white;"
        style += "}"
        return style

    @staticmethod
    def openGraphicFile() -> str:
        """
        Opens a file dialog to select a graphic file and returns the file path.
        """
        home_path = QDir.homePath()
        file_name = QFileDialog.getOpenFileName(None, "Choose an image", home_path, "Images (*.png *.xpm *.jpg)")
        if file_name:
            file_info = QFileInfo(file_name)
            name = file_info.fileName()
            destination_file_path_name = LaUtils.userImagesDirPath() + name
            QFile.copy(file_name, destination_file_path_name)
            return destination_file_path_name
        return ""

    @staticmethod
    def saveFile() -> str:
        """
        Opens a file dialog to save a file and returns the file path.
        """
        home_path = QDir.homePath()
        file_name = QFileDialog.getSaveFileName(None, "Choose a file name", LaUtils.userConversionTablesDirPath(), "*.csv")
        if file_name:
            file_info = QFileInfo(file_name)
            name = file_info.fileName()
            destination_file_path_name = LaUtils.userConversionTablesDirPath() + name
            return destination_file_path_name
        return ""